src/transport/net.cc

```diff
@@ -64,6 +64,8 @@ static_assert(sizeof(ncclNetHandle_t) <= CONNECT_SIZE, "NET Connect info is too
     } \
 } while (0);
 
+__thread connectionMetaData_t connectionMetaData;
```

Comment:
AFAIU, the call chain (within a single proxy thread) goes like so: `sendProxyConnect -> ... -> ncclNetSocketConnect -> ncclExportFlow`. Similarly, on the receiving side: `recvProxyConnect -> ... -> ncclNetSocketAccept -> ncclExportFlow`. We set `connectionMetaData` at the top of that chain, and use it within `ncclExportFlow`. I.e., there's no opportunity for it to be overwritten before it's used. Right?

If you are asking about multiple proxy threads doing the above call chains in parallel, that shouldn't be an issue because this is a thread-local var (`__thread`).

Reply:
Does a single thread deal with both `send` and `recv` for a GPU? If so, how does it guarantee that the `connectionMetaData` will not be overwritten before the switch plugin has finished using it?

Reply:
Yeah, I believe each thread handles both send and receive for its GPU.

The `connectionMetaData` struct is not shared as is with the plugin. You can see in flow_export.cc line 129 that we call the plugin's exportFlow() method with a Flow_v1 object. That object is allocated on the stack (line 100). The plugin is expected to copy that info to wherever it wants, before returning. That's the contract.